### DISCUSSION TITLE
Problems connecting to MS SQL Server 2019 on Docker container

### DISCUSSION THREAD

**Discussion by cyanezf** [mkleehammer/pyodbc#1185] [Q&A]
Please first make sure you have looked at:

* Documentation: https://github.com/mkleehammer/pyodbc/wiki - I read a lot of infor about pyodbc, unixODBC, odbcinst

### Environment

To diagnose, we usually need to know the following, including version numbers.  On Windows, be
sure to specify 32-bit Python or 64-bit:

- Python: 3.11.2 64-bit
- pyodbc: 4.0.35
- OS: Python on Windows 10 Pro x64
- DB: MSSQL Server 2019 on docker container - image: mssql/server:2019-latest
- driver: ODBC Driver 17 for SQL Server

- On Container: unixODBC 2.3.11

### Issue
I have the docker container in Synology NAS (AMD 64), IP, example: 192.169.0.3 port for container 2443

1. When I try to connect with driver:

import pyodbc
conn = pyodbc.connect('DRIVER={ODBC Driver 17 for SQL Server};SERVER=192.169.0.3,2443;UID=SA;PWD=mypassword')

Error:
Traceback (most recent call last):
  File "c:\Users\vm\Desktop\SQL Server\sql-connection.py", line 10, in <module>
    connection = odbc.connect('DRIVER={ODBC Driver 17 for SQL Server};SERVER=192.169.0.3,2443;UID=SA;PWD=mypassword')
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pyodbc.OperationalError: ('08001', '[08001] [Microsoft][ODBC Driver 17 for SQL Server]TCP Provider: The wait operation timed out.\r\n (258) (SQLDriverConnect); [08001] [Microsoft][ODBC Driver 17 for SQL Server]Login timeout expired (0); [08001] [Microsoft][ODBC Driver 17 for SQL Server]A network-related or instance-specific error has occurred while establishing a connection to SQL Server. Server is not found or not accessible. Check if instance name is correct and if SQL Server is configured to allow remote connections. For more information see SQL Server Books Online. (258)')

2. When I try to connect with DSN (DSN is configured in container, MSSQLServerDatabase is the name of DSN):

import pyodbc
conn = pyodbc.connect('DSN=MSSQLServerDatabase;UID=SA;PWD=mypassword')

Error:
Traceback (most recent call last):
  File "c:\Users\vm\Desktop\SQL Server\sql-connection.py", line 9, in <module>
    connection = odbc.connect('DSN=MSSQLServerDatabase;UID=SA;PWD=mypassword')
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pyodbc.InterfaceError: ('IM002', '[IM002] [Microsoft][ODBC Driver Manager] Data source name not found and no default driver specified (0) (SQLDriverConnect)')

How you can see, the odbc cannot connect, but when I connect to database with VSCode or SQL Server Management Studio I connect without problems.

Please, can you help me?

Thanks in advance!

**Comment by gordthompson** [comment]
So when you connect using SSMS does your connection information look exactly like this?

![Capture](https://user-images.githubusercontent.com/15894006/226144712-ac587bd9-265a-4de8-9f8e-0c9b2cbcb11a.PNG)

And are you running your Python code on the same machine that you are running SSMS?

**Reply by cyanezf** [reply]
Hello, :)

Yes, the test environment is the following:

Server side:
- 192.169.0.3 - Synology NAS 9213+ (AMD 64)
- Docker v20.10.3-1308 - Container: sql2 - mssql/server:2019-latest, for more info, If you need, I can add the docker configuration, the unixodbc config.

Client side: Laptop
- 192.169.0.20 - Dell laptop
- VMWare - SSMS, python 3.

How I commented, SSMS and VSCode can connect without problems

Thanks for your answer.

**Reply by gordthompson** [reply]
1. The Docker container does not require unixODBC. An ODBC Driver Manager (DM) like unixODBC is only used on the client side, and only for non-Windows machines. (Windows provides its own DM, so we don't need to install one.)
2. You mention VMware on your laptop. Are you running both SSMS and your Python code in the same VMware guest machine?

**Comment by cyanezf** [comment]
Hello,

1. Yes, unixODBC is inside of sql container.
2. Yes, SSMS and python are inside of VMWare guest machine, also I tested in host machine, same problem, even I tried in other computer, same problem.

**Reply by gordthompson** [reply]
> SSMS and python are inside of VMWare guest machine, also I tested in host machine, same problem, even I tried in other computer, same problem.

And in all cases you can connect using SSMS but you cannot connect using ODBC?

**Reply by cyanezf** [reply]
Exactly, SSMS works in all cases, ODBC not, so I don't know if there is some wrong in odbc.ini or odbcinst.ini files inside sql container, but I followed step by step from https://github.com/mkleehammer/pyodbc/wiki/Connecting-to-SQL-Server-from-Linux, in root and sql2 users.

**Reply by gordthompson** [reply]
The odbcinst.ini and odbc.ini files on the server (container) are not relevant when trying to connect from some other machine. They would only be used if you were running pyodbc code in the container.

On your client machine, launch

C:\Windows\System32\odbcad32.exe

and try to create an ODBC DSN for your server using "ODBC Driver 17 for SQL Server". At the end of the process, use the "Test Data Source…" button to check it. Does that work?

**Reply by cyanezf** [reply]
Hello again,

Following your guide, now I can connect to SQL server by DSN, "Test Data Source..." is ok!

So, for the python code:

import pyodbc
conn = pyodbc.connect('DSN=MSSQLServerDatabase;UID=SA;PWD=mypassword')

Works fine, thanks a lot! :)

Now, for the python code, the problem is present:

import pyodbc
conn = pyodbc.connect('DRIVER={ODBC Driver 17 for SQL Server};SERVER=192.169.0.3,2443;UID=SA;PWD=mypassword')

Error:
Traceback (most recent call last):
File "c:\Users\vm\Desktop\SQL Server\sql-connection.py", line 10, in
connection = odbc.connect('DRIVER={ODBC Driver 17 for SQL Server};SERVER=192.169.0.3,2443;UID=SA;PWD=mypassword')
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pyodbc.OperationalError: ('08001', '[08001] [Microsoft][ODBC Driver 17 for SQL Server]TCP Provider: The wait operation timed out.\r\n (258) (SQLDriverConnect); [08001] [Microsoft][ODBC Driver 17 for SQL Server]Login timeout expired (0); [08001] [Microsoft][ODBC Driver 17 for SQL Server]A network-related or instance-specific error has occurred while establishing a connection to SQL Server. Server is not found or not accessible. Check if instance name is correct and if SQL Server is configured to allow remote connections. For more information see SQL Server Books Online. (258)')

**Reply by gordthompson** [reply]
You could try using [dump_dsn](https://github.com/gordthompson/dump_dsn) to inspect the DSN entry and see if it contains something different from what you expect.

**Comment by cyanezf** [comment]
Hi,

I cannot see your comment in the web, so I answer you here.

Yes, the DSN connection conn = pyodbc.connect('DSN=my_dsn;UID=SA;PWD=mypassword') is tested and works fine in python.

The other connection conn = pyodbc.connect('DRIVER={ODBC Driver 17 for SQL Server};SERVER=192.169.0.3,2443;UID=SA;PWD=mypassword') doesn't work yet!

**Reply by gordthompson** [reply]
Sorry. I deleted my comment because I realized that you had already answered that question.
